internal/profiling/profiling.go

```diff
@@ -0,0 +1,220 @@
+//
+// Note that the circular buffer simply takes any interface{}. In the future,
+// more types of measurements (such as the number of memory allocations) could
+// be measured, which might require a different type of object being pushed
+// into the circular buffer.
+
+package profiling
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// 0 or 1 representing profiling off and on, respectively. Use IsEnabled and
+// SetEnabled to get and set this in a safe manner.
+var profilingEnabled uint32
+
+// IsEnabled returns whether or not profiling is enabled.
+func IsEnabled() bool {
+	return atomic.LoadUint32(&profilingEnabled) > 0
+}
+
+// Enable turns profiling on and off.
+//
+// Note that it is impossible to enable profiling for one server and leave it
+// turned off for another. This is intentional and by design -- if the status
+// of profiling was server-specific, clients wouldn't be able to profile
+// themselves. As a result, SetEnabled turns profiling on and off for all
+// servers and clients in the binary. Each stat will be, however, tagged with
+// whether it's a client stat or a server stat; so you should be able to filter
+// for the right type of stats in post-processing.
+// SetEnabled is the internal
+func Enable(enabled bool) {
+	if enabled {
+		atomic.StoreUint32(&profilingEnabled, 1)
+	} else {
+		atomic.StoreUint32(&profilingEnabled, 0)
+	}
+}
+
+// A Timer represents the wall-clock beginning and ending of a logical
+// operation.
+type Timer struct {
+	TimerTag string
+	Begin    time.Time
+	End      time.Time
+	GoID     int64
+}
+
+// NewTimer creates and returns a new Timer object. This is useful when you
+// don't already have a Stat object to associate this Timer with; for example,
+// before the context of a new RPC query is created, a Timer may be needed to
+// measure transport-related operations.
+//
+// Use AppendTimer to append the returned Timer to a Stat.
+func NewTimer(timerTag string) *Timer {
+	return &Timer{
+		TimerTag: timerTag,
+		Begin:    time.Now(),
+		GoID:     goid(),
+	}
+}
+
+// A Stat is a collection of Timers that represent timing information for
+// different components within this Stat. For example, a Stat may be used to
+// reference the entire lifetime of an RPC request, with Timers within it
+// representing different components such as encoding, compression, and
+// transport.
+//
+// The user is expected to use the included helper functions to do operations
+// on the Stat such as creating or appending a new timer. Direct operations on
+// the Stat's exported fields (which are exported for encoding reasons) may
+// lead to data races.
+type Stat struct {
+	StatTag  string
+	Timers   []Timer
+	Metadata []byte
+	mu       sync.Mutex
+}
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation.
+const defaultStatAllocatedTimers int32 = 128
```

Comment:
What is the rationale behind the value of 128?

Reply:
A typical unary request creates about 80-100 timers. This is mostly anecdotal and may vary depending on different factors, but this is the average number I've seen in the past few weeks. Should I include this in the comment?

Reply:
Sure, a comment should be helpful for the future.